modin/pandas/test/test_groupby.py

```diff
@@ -1731,6 +1785,168 @@ def test_not_str_by(by, as_index):
+        pytest.param(
+            {1: "sum", 2: "nunique", -1: "nunique"},
+            id="dict_agg_has_intersection_with_categorical_by",
+        ),
+    ],
+)
+# There are two versions of the `handle_as_index` method: the one accepting pandas.DataFrame from
+# the execution kernel and backend agnostic. This parameter indicates which one implementation to use.
+@pytest.mark.parametrize("use_backend_agnostic_method", [True, False])
+def test_handle_as_index(
+    internal_by_length,
+    external_by_length,
+    has_categorical_by,
+    agg_func,
+    use_backend_agnostic_method,
+    request,
+):
+    """
+    Test ``modin.core.dataframe.algebra.default2pandas.groupby.GroupBy.handle_as_index``.
+
+    The role of the ``handle_as_index`` method is to build a groupby result considering
+    ``as_index=False`` from the result that was computed with ``as_index=True``.
+
+    So the testing flow is the following:
+        1. Compute GroupBy result with the ``as_index=True`` parameter via Modin.
+        2. Build ``as_index=False`` result from the ``as_index=True`` using ``handle_as_index`` method.
+        3. Compute GroupBy result with the ``as_index=False`` parameter via pandas as the reference result.
+        4. Compare the result from the second step with the reference.
+    """
+    by_length = internal_by_length + external_by_length
+    if by_length == 0:
+        pytest.skip("No keys to group on were passed, skipping the test.")
+
+    if (
+        has_categorical_by
+        and by_length > 1
+        and (
+            isinstance(agg_func, dict)
+            or ("nunique" in request.node.callspec.id.split("-"))
+        )
+    ):
+        pytest.skip(
+            "The linked bug makes pandas raise an exception when 'by' is categorical: "
+            "https://github.com/pandas-dev/pandas/issues/36698"
+        )
+
+    df = pandas.DataFrame(test_groupby_data)
+    external_by_cols = GroupBy.validate_by(df.add_prefix("external_"))
+
+    if has_categorical_by:
+        df = df.astype({df.columns[-1]: "category"})
+
+    if isinstance(agg_func, dict):
+        agg_func = {df.columns[key]: value for key, value in agg_func.items()}
+        selection = list(agg_func.keys())
+        agg_dict = agg_func
+        agg_func = lambda grp: grp.agg(agg_dict)  # noqa: E731 (lambda assignment)
+    else:
+        selection = None
+
+    # Selecting 'by' columns from both sides of the frame so they located in different partitions
+    internal_by = df.columns[
+        range(-internal_by_length // 2, internal_by_length // 2)
+    ].tolist()
+    external_by = external_by_cols[:external_by_length]
+
+    pd_by = internal_by + external_by
+    md_by = internal_by + [pd.Series(ser) for ser in external_by]
+
+    grp_result = pd.DataFrame(df).groupby(md_by, as_index=True)
+    grp_reference = df.groupby(pd_by, as_index=False)
+
+    agg_result = agg_func(grp_result)
+    agg_reference = agg_func(grp_reference)
+
+    if use_backend_agnostic_method:
```

Comment:
I would prefer to change all entries having `*backend*`.

Reply:
Created an issue for this (#3787). Will fix soon

Reply:
Thanks!